internal/handler/cmd_query.go

```diff
@@ -19,18 +19,44 @@ import (
 	"fmt"
 	"strings"
 
+	"github.com/FerretDB/FerretDB/internal/clientconn/conninfo"
 	"github.com/FerretDB/FerretDB/internal/handler/common"
 	"github.com/FerretDB/FerretDB/internal/handler/handlererrors"
 	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
 	"github.com/FerretDB/FerretDB/internal/util/must"
 	"github.com/FerretDB/FerretDB/internal/wire"
 )
 
 // CmdQuery implements deprecated OP_QUERY message handling.
 func (h *Handler) CmdQuery(ctx context.Context, query *wire.OpQuery) (*wire.OpReply, error) {
-	cmd := query.Query().Command()
+	q := query.Query()
+	cmd := q.Command()
 	collection := query.FullCollectionName
 
+	v, _ := q.Get("speculativeAuthenticate")
+	if v != nil && (cmd == "ismaster" || cmd == "isMaster" || cmd == "hello") {
```

Comment:
@b1ron what driver or situation `hello` is used for speculative authenticate? Is using response of `IsMaster` correct for `hello` command as well?

Reply:
> If `speculativeAuthenticate` succeeds, it skips `saslStart`.

Where did you get this information from?

Based on my research and testing `speculativeAuthenticate` is used to reduce connection overhead time by embedding the body of the `saslStart` command into the command `hello` under the field `speculativeAuthenticate`. No steps are skipped AFAIK. After further testing it can also be sent in an OP_MSG, see below packet.

> what driver or situation `hello` is used for speculative authenticate

I am unsure of every situation this can occur. A packet from the legacy shell which is using an OP_MSG: 

Note I am connecting to MongoDB here.

```
Frame 16: 532 bytes on wire (4256 bits), 532 bytes captured (4256 bits) on interface lo, id 0
Ethernet II, Src: 00:00:00_00:00:00 (00:00:00:00:00:00), Dst: 00:00:00_00:00:00 (00:00:00:00:00:00)
Internet Protocol Version 4, Src: 127.0.0.1, Dst: 127.0.0.1
Transmission Control Protocol, Src Port: 35596, Dst Port: 27017, Seq: 1, Ack: 1, Len: 466
Mongo Wire Protocol
    Message Length: 466
    Request ID: 0x00000000 (0)
    Response To: 0x00000000 (0)
    OpCode: Extensible Message Format (2013)
    Message Flags: 0x00000001, ChecksumPresent
    Section
        Kind: Body (0)
        BodyDocument
            Document length: 441
            Elements
                Element: hello
                Element: speculativeAuthenticate
                    Type: Document (0x03)
                    Document
                        Document length: 122
                        Elements
                            Element: saslStart
                            Element: mechanism
                            Element: payload
                            Element: db
                Element: saslSupportedMechs
                Element: client
                Element: $db
[Malformed Packet: MONGO]
```

> Is using response of IsMaster correct for hello command as well?

The `hello` and `isMaster` responses are nearly identical, so I think either will work  given that we continue the SCRAM conversation.